### ISSUE TITLE
Incorrect suffix retained after election

### ISSUE THREAD

**Issue by eddyashton** [microsoft/CCF#2803]
The changes in #2801 have hit a Raft bug very similar to the one described in #2189 (and believed fixed in #2192). Specifically, that fix to "accept a rollback instruction from the new primary" is not reached if the node rejects the AppendEntries early in processing, for instance because it is not contiguous with its current ledger or it disagrees on the previous term. The former is legitimate, but the latter is not and can lead to permanent commit lock.

From spelunking in some logs, one incorrect flow was observed:
- Node 0 is briefly primary in term 8
- Node 0 produces a signature at 8.25, but doesn't manage to tell anyone about it
- Node 1 becomes primary in term 9
- Node 1 produces a transaction at 9.25, and proceeds to 9.26
- Node 1 sends AppendEntries to 0 containing 9.25 -> 9.26
- Node 0 sees an AppendEntries in term 9, and dutifully rolls back to its `last_committable` (setting `is_new_follower`, but hoping that its 25 is still useful)
- Node 0 processes the AppendEntries and says "no, that looks like junk to me"

0's logs show it rejecting an AppendEntries it should have accepted:
```
24:49.243 |00| Received append entries: 9.25 to 9.26 (from 1=bf9b in term 9) (raft.h:1350)
24:49.243 |00| Previous term for 25 should be 8 (raft.h:1439)
24:49.243 |00| Recv append entries to n[0=92bf] from n[1=bf9b] but our log at 25 has the wrong previous term (ours: 8, theirs: 9) (raft.h:1461)
24:49.243 |00| Send append entries response from 0=92bf to 1=bf9b for index 25: 1 (raft.h:2084)
```

This is a more general problem than that precise flow. For instance the locked node could have been any follower of 0 in term 8. Its not critical that they produced the signature, just that it is the final entry in their log and disagrees with an AppendEntries they should be obeying.

I'm currently considering 2 possible fixes:
- Obey an AppendEntries rollback in more circumstances. This should just involve moving a block of code earlier in the `recv_append_entries` function. However reasoning about exactly where it should go is tricky, as we want to (if possible) keep the current behaviour that it is robust to replay (we only do this immediately after becoming follower, if we did it blindly then a malicious host could trivially cause us to repeatedly roll back to this point and lose state).
- Change what we report to the primary when rejecting AppendEntries. The Raft paper says we should return the first TxID we disagree with, so the host knows where the next AppendEntries should start. We actually send our latest index (which may be junk that we're holding optimistically while in `is_new_follower`), and the primary treats that as the matching index. Sending the last we agree with should be equivalent to the first we disagree with, and may be a small fix that fixes everything.

I still don't know why this has arisen from #2801, and will look for a simpler repo for a standalone fix.

**Comment by eddyashton** [comment]
After discussion with @achamayou and @jumaffre, we have a rough plan for a fix.

One point of discussion was whether, as a new follower with an optimistically retained suffix, you should advertise a _late_ or _early_ index when responding to AppendEntries.
- It may be possible to respond with the latest index you think may be good, and then you may backtrack through multiple steps with the primary as they send you new suffices you still disagree with. For this to work, the primary would have to distinguish those responses, and not treat them as a matching index. With careful design, this could be handled as a single rollback (ie - retaining a single multi-term suffix while backtracking through each of the mismatching terms in it, and only rolling back once we find a confirmed match).
- Alternatively a node can respond with an early estimate of the matching index - their current `commit_idx`. When processing the AppendEntries sent as a result of this, we'd expect the initial entries to be trivially accepted (from a known term at a known index), and eventually to reach an entry which indicates a divergence (still in the `is_new_follower` state, spotted a term increment, the new entry disputes something in our current ledger => rollback to here and apply the new entries). A result of this approach is that the advertised indexes from a majority of nodes may temporarily move backwards, as far as this protocol is concerned. As long as the primary uses _their_ election index as a lower-bound for re-establishing commit index, and does not look for a simple majority in these responses, then this does not allow the commit index to move backwards. In the pessimal case, this could result in replaying a long, redundant suffix that the receiving node already has, but its length should be reasonable in a healthy network (and perhaps trigger a general "too out-of-date for AppendEntries catchup, use a snapshot instead" behaviour in future).

We believe the latter is a better approach, and I'll look at implementing that. Still contingent on finding a reliable minimal repro.

**Comment by eddyashton** [comment]
After some further investigation I've worked out why this started happening with the change to host management of node-to-node connections, and have a repro to test the fix against.

In our old channel management, we try to re-use an existing TCP connection for as long as possible. This means that in the partitions test, once a partition is healed, we actually reconnect an existing connection and try to transmit new messages on it. This means that all _old_ messages, which were sent during the partition, are still queued and delivered correctly (since TCP provides reliable, ordered transmission, if we insist on using the same connection it will only deliver the newer ones once it has successfully re-delivered the old). So the reconnected node still sees the first AppendEntries from the new primary, which has the special property that its prev-tx_id agrees with the reconnected node (it points at the last transaction they both hold). This is processed correctly, resulting in a rollback to that index, and the new suffix being applied as it arrives. However, if this single message is dropped, the reconnecting node is stuck - every later AppendEntries starts from a higher tx_id, which it incorrectly disagrees with, and it refuses to rejoin the quorum.

**Comment by eddyashton** [comment]
There was a partial fix for this in #2835, but I believe it has introduced another issue that could lead to loss of persistence in some situations. To go back to the discussion above about advertising _late_ or _early_ indices in a NACK, the idea of trivially ignoring a matching prefix of an AppendEntries is actually quite hard to do in our implementation. We currently rollback when an AppendEntries is received, so if it does not contain the full matching prefix we may have lost persistence. And even if we move the rollback to during the processing of each entry, the comparison would be "does the TxID of this log entry in this AppendEntries match the TxID of the same index in my current log", and we do not know the TxID of each log entry as we examine it (only when we see the next signature, or the end of the AppendEntries). So the recognition of a conflict needs to happen after parsing and applying an entry, we cannot do it (as desired) in advance.

Instead, I think we need to do the _late_ estimation and back-tracking approach. I've currently convinced myself that the matching index is derivable systematically, but will test some more edge cases. Consider a situation where 2 nodes are past the commit index in term N, so either is leader-eligible. Node A becomes the leader, but makes no commit progress, then Node B becomes leader and must find the end of A's matching ledger. This last matching entry is in term N, and is the smaller of the 2 node's opinions of when term N ended.

**Comment by eddyashton** [comment]
A detailed scenario that I've used as the basis for a unit test to confirm this behaviour and a potential fix:

  - Produce an initial state where A is primary in term 1 of a 5-node network,
    has mixed success replicating its entries. Local commit index is marked by
    []. True commit index is the highest of these, [1.2].
```
    A:  1.1 [1.2] 1.3
    B: [1.1] 1.2  1.3
    C: [1.1] 1.2
    D: [1.1]
    E: [1.1]
```

  - Intuitively, B and C are responsible for persisting 1.2, although they don't
    'know' this locally.

  - Node A dies. Node B becomes primary, and creates some tail junk that no-one
    else sees. It makes no commit progress. Crucially this is committable (as
    all of these indices are), so B is reluctant to discard the entries in term
    2, and 1.3, though all _could_ be discarded.
```
    A:  1.1 [1.2] 1.3 (DEAD)
    B: [1.1] 1.2  1.3  2.4  2.5
    C: [1.1] 1.2
    D: [1.1]
    E: [1.1]
```

  - Node C doesn't hear from B for long enough, and becomes primary. Its
    AppendEntries are regularly lost. Eventually a heartbeat ("I'm at 1.2")
    reaches Node B, which responds with a NACK ("I don't accept that tail").
    Node C responds to that NACK with a better AppendEntries, which _should_
    bring Node B back to the committed level, and forward to wherever C is.
    However, under the current implementation, this AppendEntries may
    actually cause B to roll back further than is safe, losing the committed
    state.
```
    C->B: AE [1.2, 1.2)
    B->C: AER NACK 1.1 (B's commit index)
    C->B: AE [1.1, 1.1) (Large entries mean this is only a partial catchup)
    B: Rolls back to 1.1

    B: [1.1]
    C: [1.1] 1.2
    D: [1.1]
    E: [1.1]
```
  - At this point we have lost persistence; a committed index (1.2) is no
    longer present on a majority of nodes. While C is unlikely to advertise
    it (fancy election rules mean they're waiting for commit at 1.2) and
    will continue to share it, it's possible for C to die here and B, D, or
    E to win an election and proceed without this committed suffix,
    forking/overwriting 1.2 with 4.2.
 
It may also possible to hit this with 3 nodes, where the 3rd just votes for the others to win alternating elections and advance their non-committed suffix, and only after a long divergence catches up to allow one to advance their commit index? I'll try to represent that in a test case as well for complete coverage.
